kbkdf/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+#![no_std]
+#![doc = include_str!("../README.md")]
+
+use core::{fmt, marker::PhantomData, ops::Mul};
+use digest::{
+    array::{typenum::Unsigned, Array, ArraySize},
+    consts::{U32, U8},
+    crypto_common::KeySizeUser,
+    typenum::op,
+    KeyInit, Mac,
+};
+
+pub mod sealed;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    InvalidRequestSize,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidRequestSize => write!(
+                f,
+                "Request output size is too large for the value of R specified"
+            ),
+        }
+    }
+}
+
+impl core::error::Error for Error {}
+
+// Helper structure along with [`KbkdfUser`] to compute values of L and H.
+struct KbkdfCore<OutputLen, PrfOutputLen> {
+    _marker: PhantomData<(OutputLen, PrfOutputLen)>,
+}
+
+trait KbkdfUser {
+    // L - An integer specifying the requested length (in bits) of the derived keying material
+    // KOUT. L is represented as a bit string when it is an input to a key-derivation function. The
+    // length of the bit string is specified by the encoding method for the input data.
+    type L;
+
+    // h - An integer that indicates the length (in bits) of the output of a single invocation of the
+    // PRF.
+    type H;
+}
+
+impl<OutputLen, PrfOutputLen> KbkdfUser for KbkdfCore<OutputLen, PrfOutputLen>
+where
+    OutputLen: ArraySize + Mul<U8>,
+    <OutputLen as Mul<U8>>::Output: Unsigned,
+    PrfOutputLen: ArraySize + Mul<U8>,
+    <PrfOutputLen as Mul<U8>>::Output: Unsigned,
+{
+    type L = op!(OutputLen * U8);
+    type H = op!(PrfOutputLen * U8);
+}
+
+/// [`Kbkdf`] is a trait representing a mode of KBKDF.
+/// It takes multiple arguments:
+///  - Prf - the Pseudorandom Function to derive keys from
+///  - K - the expected output length of the newly derived key
+///  - R - An integer (1 <= r <= 32) that indicates the length of the binary encoding of the counter i
+///        as an integer in the interval [1, 2r − 1].
+pub trait Kbkdf<Prf, K, R: sealed::R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArraySize + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArraySize + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+{
+    /// Derives `key` from `kin` and other parameters.
+    fn derive(
+        &self,
+        kin: &[u8],
+        use_l: bool,
+        use_separator: bool,
+        use_counter: bool,
```

Comment:
API of this function is somewhat easy to misuse. Maybe it's worth to do something like this?
```rust
let key = k.derive(Params {
    secret,
    label: b"label",
    context: b"ctx",
    use_l: true,
    ..Default::default(),
});
```

Reply:
We can't provide a default for the boolean as they would usually default to true and can't provide a default for the key, label or context, but we can make a param builder: https://github.com/TheBestTvarynka/KDFs/pull/8

Reply:
done